packages/babel-traverse/src/path/removal.ts

```diff
@@ -23,8 +23,34 @@ export function remove(this: NodePath) {
 }
 
 export function _removeFromScope(this: NodePath) {
-  const bindings = this.getBindingIdentifiers();
```

Comment:
I think ideally this issue should have been fixed in `getBindingIdentifiers`. An assignment expression does not introduce new bindings and therefore it should not contain any binding identifiers: In the spec the AssignmentPattern production is resolved to IdentifierReference:

https://tc39.es/ecma262/#sec-destructuring-assignment

However, practically the `getBindingIdentifiers` is used to retrieve identifier references / binding identifiers in a destructuring pattern so unless we come up with a `getBindingAndReferenceIdentifiers`, we may have to live with that.

The assignment expression is registered as a constant violation in the binding info, so to fix this issue, we can check whether `this` is an AssignmentExpression, if so, 
instead of removing the binding, we should check the binding info and remove the registered `constantViolations`.

And we should also add a new test case for ForXStatement:

```js
var x = 1;
for (x of []) {} // <-- removing this should not remove the binding
```

Reply:
Thanks for your great point! It should work now!